scripts/secureboot-certs

```diff
@@ -1,160 +1,577 @@
-
-uefi_pca_pem=uefi_pca.pem
-openssl x509 -in ${pca} -inform DER -outform PEM -out ${uefi_pca_pem}
-
-uefi_kek_pem=uefi_kek.pem
-openssl x509 -in ${kek} -inform DER -outform PEM -out ${uefi_kek_pem}
-
-# Generate KEK
-openssl req -newkey rsa:4096 -nodes -new -x509 -sha256 -days 3650 \
-        -subj "/CN=KEK Owner/" \
-        -keyout KEK.key \
-        -out KEK.crt
-
-# Self-sign KEK and create KEK.auth
-# NOTE: because this cert is self-signed it will not be provisionable from
-# the guest, only as a file for uefistored to use from /var/lib/uefistored
-/opt/xensource/libexec/create-auth -k KEK.key -c KEK.crt KEK KEK.auth KEK.crt ${uefi_kek_pem}
-
-# Sign Microsoft certs with your KEK and create a db.auth
-/opt/xensource/libexec/create-auth \
-    -k KEK.key \
-    -c KEK.crt \
-    db \
-    db.auth \
-    ${uefi_pca_pem} \
-    ${uefi_ca_pem}
-
-install -m 444 db.auth ${AUTH_DIR}/db.auth
-install -m 444 KEK.auth ${AUTH_DIR}/KEK.auth
-
-# If you would like the dbx, it may be found here:
-# https://www.uefi.org/revocationlistfile
-#
-# It is already in .auth format, so simply download it to
-#  /var/lib/uefistored/dbx.auth
-
-popd
-rm -R ${tmpdir}
+#!/usr/bin/env python
+"""
+This script installs UEFI certificates for XCP-ng hosts.
+"""
+import argparse
+import atexit
+import base64
+import hashlib
+import logging
+import os
+import requests
+import shutil
+import subprocess
+import sys
+import tarfile
+import tempfile
+
+from io import BytesIO
+
+import XenAPI
+
+__author__ = "Bobby Eshleman"
+__copyright__ = "Copyright 2021, Vates SAS"
+__license__ = "GPLv2"
+__version__ = "1.0.0"
+__maintainer__ = "Bobby Eshleman"
+__email__ = "bobbyeshleman@gmail.com"
+__status__ = "Production"
+
+
+TEMPDIRS = []
+
+
+def cd_tempdir():
+    """Change directories to a temporary directory.
+
+    Usage:
+
+    ```
+        prevdir = cd_tempdir()
+
+        # Do stuff inside temporary directory
+        ...
+
+        # Return to previous directory
+        os.chdir(prevdir)
+    ```
+
+    All temporary directories are automatically cleaned up upon program exit.
+
+    See function `cleanup`.
+
+    Return the name of the current directory.
+    """
+    global TEMPDIRS
+    prevdir = os.path.abspath(os.curdir)
+    tempdir = tempfile.mkdtemp()
+    os.chdir(tempdir)
+    TEMPDIRS.append(tempdir)
+    return prevdir
+
+
+def cleanup():
+    global TEMPDIRS
+    for tempdir in TEMPDIRS:
+        shutil.rmtree(tempdir)
+
+
+class Scopes:
+    SYSTEM = "system"
+    POOL = "pool"
+
+
+class Actions:
+    CLEAR = "clear"
+    INSTALL = "install"
+    REPORT = "report"
+
+
+class Urls:
+    CA = "https://www.microsoft.com/pkiops/certs/MicCorUEFCA2011_2011-06-27.crt"
+    PCA = "https://www.microsoft.com/pkiops/certs/MicWinProPCA2011_2011-10-19.crt"
+    KEK = "https://www.microsoft.com/pkiops/certs/MicCorKEKCA2011_2011-06-24.crt"
+    dbx = "https://uefi.org/sites/default/files/resources/dbxupdate_x64.bin"
```

Comment:
Why is `dbx` not capped when the other Urls are?

Reply:
For some reason in the UEFI spec `db` and `dbx` are always lower case and `KEK` and others are always upper case.   I can only suspect that it is because `db`/`dbx` are not "true" abbreviations (db stands for "Image Security Database" so really should be `isb`, while KEK stands for "Key Exchange Key".). I've tried to retain their pattern where possible.